lib/shard/src/segment_holder/mod.rs

```diff
@@ -96,6 +96,24 @@ impl SegmentHolder {
             .chain(self.non_appendable_segments.iter())
     }
 
+    /// Get filesystem aligned UUIDs instead of internal segment IDs for reporting in telemetry
+    pub fn segment_uuids(&self, segment_ids: &[SegmentId]) -> Vec<String> {
+        segment_ids
+            .iter()
+            .filter_map(|id| {
+                self.appendable_segments
+                    .get(id)
+                    .or_else(|| self.non_appendable_segments.get(id))
+                    .map(|segment| segment.get().read().segment_uuid())
+                    .or_else(|| {
+                        // Shouldn't happen
+                        log::warn!("Failed to get UUID for segment with internal id {id}");
+                        None
+                    })
+            })
+            .collect::<Vec<_>>()
+    }
```

Comment:
Nit: I'd rather inline this function into the caller, e.g.:

```rust
let nonoptimal_segment_uuids = {
    let segments = segments.read();
    nonoptimal_segment_ids
        .iter()
        .filter_map(|id| segments.get(*id))
        .map(|segment| segment.get().read().segment_uuid())
        .collect::<Vec<_>>()
};
```

Note: with a small chance it is possible that `segments.get()` will return `None` because the segment can removed while we're not holding the lock.